packages/block-editor/src/components/editor-styles/index.js

```diff
@@ -67,11 +74,42 @@ function useDarkThemeBodyClassName( styles ) {
 	);
 }
 
-export default function EditorStyles( { styles } ) {
-	const stylesArray = useMemo(
-		() => Object.values( styles ?? [] ),
-		[ styles ]
-	);
+export default function EditorStyles( { styles, children } ) {
+	const [ overrides, setOverrides ] = useState( [] );
+	const updateStyle = useCallback( ( style ) => {
+		setOverrides( ( _overrides ) => {
+			const index = _overrides.findIndex(
+				( override ) => override.id === style.id
+			);
+			if ( index === -1 ) {
+				return [ ..._overrides, style ];
+			}
+			return [
+				..._overrides.slice( 0, index ),
+				style,
+				..._overrides.slice( index + 1 ),
+			];
+		} );
+		return () => {
+			setOverrides( ( _overrides ) =>
+				_overrides.filter( ( override ) => override.id !== style.id )
+			);
+		};
+	}, [] );
+	const stylesArray = useMemo( () => {
+		const _styles = Object.values( styles ?? [] );
+
+		for ( const override of overrides ) {
+			const index = _styles.findIndex( ( { id } ) => id === override.id );
+			if ( index === -1 ) {
+				_styles.push( override );
+			} else {
+				_styles[ index ] = override;
+			}
+		}
+
+		return _styles;
+	}, [ styles, overrides ] );
```

Comment:
Super early feedback: I think all these style/svg transformations and iterations could be absorbed in a single `useMemo`.